llama-index-integrations/llms/llama-index-llms-alephalpha/llama_index/llms/alephalpha/base.py

```diff
@@ -0,0 +1,246 @@
+            context_window=alephalpha_modelname_to_contextsize(self.model),
+            num_output=self.max_tokens,
+            is_chat_model=False,  # The Aleph Alpha API does not support chat yet
+            model_name=self.model,
+        )
+
+    @property
+    def tokenizer(self) -> Tokenizer:
+        client = self._get_client()
+        return client.tokenizer(model=self.model)
+
+    @property
+    def _model_kwargs(self) -> Dict[str, Any]:
+        base_kwargs = {
+            "model": self.model,
+            "temperature": self.temperature,
+            "maximum_tokens": self.max_tokens,
+        }
+        return {
+            **base_kwargs,
+            **self.additional_kwargs,
+        }
+
+    @property
+    def _completion_kwargs(self) -> Dict[str, Any]:
+        base_kwargs = {
+            "maximum_tokens": self.max_tokens,
+            "temperature": self.temperature,
+            "repetition_penalties_include_prompt": self.repetition_penalties_include_prompt,
+            "repetition_penalties_include_completion": self.repetition_penalties_include_completion,
+            "sequence_penalty": self.sequence_penalty,
+            "sequence_penalty_min_length": self.sequence_penalty_min_length,
+            "stop_sequences": self.stop_sequences,
+        }
+        return {**base_kwargs}
+
+    def _get_all_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
+        return {
+            **self._model_kwargs,
+            **kwargs,
+        }
+
+    def _get_credential_kwargs(self) -> Dict[str, Any]:
+        return {
+            "token": self.token,
+            "host": self.base_url,
+            "hosting": self.hosting,
+            "request_timeout_seconds": self.timeout,
+            "total_retries": self.max_retries,
+            "nice": self.nice,
+            "verify_ssl": self.verify_ssl,
+        }
+
+    def _get_client(self) -> Client:
+        if self._client is None:
+            self._client = Client(**self._get_credential_kwargs())
+        return self._client
+
+    def _get_aclient(self) -> AsyncClient:
+        if self._aclient is None:
+            self._aclient = AsyncClient(**self._get_credential_kwargs())
+        return self._aclient
+
+    @llm_chat_callback()
+    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
+        raise NotImplementedError("Aleph Alpha does not currently support chat.")
+
+    @llm_completion_callback()
+    def complete(
+        self, prompt: str, formatted: bool = False, **kwargs: Any
+    ) -> CompletionResponse:
+        client = self._get_client()
+        all_kwargs = {"prompt": Prompt.from_text(prompt), **self._completion_kwargs}
+
+        request = CompletionRequest(**all_kwargs)
+
+        response = client.complete(request=request, model=self.model)
+        completion = response.completions[0].completion
+        return CompletionResponse(text=completion, raw=response.to_json())
+
+    @llm_completion_callback()
```

Comment:
just noting that, our `callbacks` will be replaced by our new `instrumentation`. We will support `callbacks` for a while still but just giving you a heads up that this mechanism will need to change at some point. :)